evidence.go

```diff
@@ -0,0 +1,279 @@
+	"strings"
+
+	cose "github.com/veraison/go-cose"
+	ptoken "github.com/veraison/psatoken"
+)
+
+// claimType to hold the value of claim
+
+type claimType int
+
+const (
+	PlatformClaim claimType = iota + 1
+	RealmClaim
+)
+
+type CcaToken struct {
+	CcaPlatformToken *[]byte `cbor:"44234,keyasint" json:"cca-platform-token"`
+	CcaRealmToken    *[]byte `cbor:"44241,keyasint" json:"cca-realm-delegated-token"`
+}
+
+// CcaEvidence is a wrapper around CcaToken
+type CcaEvidence struct {
+	CcaPlatformClaims ptoken.IClaims
+	CcaRealmClaims    IClaims
+	message           *CcaToken
+}
+
+func (e *CcaEvidence) SetCcaPlatformClaims(c ptoken.IClaims) error {
+	if err := c.Validate(); err != nil {
+		return fmt.Errorf("validation of cca-platform-claims failed: %w", err)
+	}
+
+	e.CcaPlatformClaims = c
+
+	return nil
+}
+
+func (e *CcaEvidence) SetCcaRealmClaims(c IClaims) error {
+	if err := c.Validate(); err != nil {
+		return fmt.Errorf("validation of cca-realm-claims failed: %w", err)
+	}
+
+	e.CcaRealmClaims = c
+	return nil
+}
+
+// Sign signs the given evidence using the supplied Platform and Realm Signer
+// and returns the complete CCA Evidence collection token as CBOR bytes
+
+func (e *CcaEvidence) Sign(pSigner cose.Signer, rSigner cose.Signer) ([]byte, error) {
+
+	if e.CcaPlatformClaims == nil {
+		return nil, fmt.Errorf("missing platform claims in evidence")
+	}
+
+	if e.CcaRealmClaims == nil {
+		return nil, fmt.Errorf("missing realm claims in evidence")
+	}
+
+	if err := e.signClaim(PlatformClaim, pSigner); err != nil {
+		return nil, err
+	}
+
+	if err := e.signClaim(RealmClaim, rSigner); err != nil {
+		return nil, err
+	}
+
+	// We do now have CcaPlatform and Realm Token setup correctly.
+	buf, err := em.Marshal(e.message)
+	if err != nil {
+		return nil, fmt.Errorf("cbor encoding of CCA token failed: %w", err)
+	}
+
+	return buf, nil
+}
+
+func (e *CcaEvidence) signClaim(cl claimType, signer cose.Signer) error {
+	var err error
+
+	message := cose.NewSign1Message()
+	if cl == PlatformClaim {
+		message.Payload, err = e.CcaPlatformClaims.ToCBOR()
+		if err != nil {
+			return err
+		}
+	} else if cl == RealmClaim {
+		message.Payload, err = e.CcaRealmClaims.ToCBOR()
+		if err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("invalid claimType %d: for signing", cl)
+	}
+
+	alg := signer.Algorithm()
+
+	if strings.Contains(alg.String(), "unknown algorithm value") {
+		return errors.New("signer has no algorithm")
+	}
+
+	message.Headers.Protected.SetAlgorithm(alg)
+
+	err = message.Sign(rand.Reader, []byte(""), signer)
+	if err != nil {
+		return err
+	}
+
+	wrap, err := message.MarshalCBOR()
+	if err != nil {
+		return fmt.Errorf("unable to MarshalCBOR %d claimType = %w", cl, err)
+	}
+	if e.message == nil {
+		ccaToken := CcaToken{}
+		e.message = &ccaToken
+	}
+
+	if cl == PlatformClaim {
+		e.message.CcaPlatformToken = &wrap
+	} else {
+		e.message.CcaRealmToken = &wrap
+	}
+	return nil
+}
+
+// Semantic validation, of CCA token
```

Comment:
there is no semantic validation here.  syntactic perhaps?

```suggestion
// Semantic validation of CCA token
```